### ISSUE TITLE
Decisions about the API

### ISSUE THREAD

**Issue by inf3rno** [inf3rno/patterns#6]
I need to decide what kind of API will be the default. Currently it is based on function calls in the example, like `word()`, `number()`, `concatenate()`, `alternatives()`, `compile()`, etc...

**Comment by inf3rno** [comment]
Just an example:

```js
var num = number();
var wrd = word();
var x = alternatives(num, wrd);
var y = alternatives(num, wrd);
var mapping = {
    x: {
        type: key(x, {n:num,w:wrd}),
        value: x
    },
    y: {
        type: key(y, {n:num, w:wrd}),
        value: y
    }
};
var p = compile(concatenate(x,y), mapping);
p.find(pointer("abc123",0));
// {x:{type: "w", value: "abc"}, y:{type:"n", value: "123"}}
```

Is it possible to do the same in a simpler way?

**Comment by inf3rno** [comment]
Doing with the #5 string based mapping is an option to have a more compact possibly one liner code for the pattern. We could do the following:

```js
var xy = concatenate(
    alternatives(
        number().as("num"), 
        word().as("wrd")
    ).as("x"),
    copy("x").as("y")
);
var mapping = {
    x: {
        type: keyForValue("x", {n:"num",w:"wrd"}),
        value: "x"
    },
    y: {
        type: keyForValue("y", {n:"num", w:"wrd"}),
        value: "y"
    }
};
var p = compile(xy, mapping);
p.find(pointer("abc123",0));
// {x:{type: "w", value: "abc"}, y:{type:"n", value: "123"}}
```

**Comment by inf3rno** [comment]
I like the one liner syntax, it is not that bad. What I don't like here is the mapping. I don't think it should be part of the compiling. It should be a next step. So the compiled pattern should return always the same output and a mapper could change that. In this case:

```js
var xy = concatenate(
    alternatives(
        number().as("num"), 
        word().as("wrd")
    ).as("x"),
    copy("x").as("y")
);
var result = pattern(xy).match("abc123");
console.log(result);
```
```js
{
    x: {wrd: "abc", toString: () => "abc"},
    y: {num: "123", toString: () => "123"},
    toString: () => "abc123"
}
```

**Comment by inf3rno** [comment]
Well this whole keyForValue thing is because we wanted to do everything in a single step. So we did not want to check the `number|word` subpattern again. What if we add the type by default instead of using keys?
```js
{
    x: {group: "wrd", toString: () => "abc"},
    y: {group: "num", toString: () => "123"},
    toString: () => "abc123"
}
```
I like this better, because we have the info we needed, still it is not right, because it returns the value with toString, and if the `wrd` group is nested too, then we need the match instead of the string representation of it. We could do this:
```js
{
    value: [
        {
            group: "x", 
            value: {
                group: "wrd",
                value: "abc",
                toString: () => "abc"
            }, 
            toString: () => "abc"
        },
        {
            group: "y", 
            value: {
                group: "num",
                value: "123",
                toString: () => "123"
            }, 
            toString: () => "123"
        }
    ],
    toString: () => "abc123"
}
```
It is a lot more verbose and contains the necessary info. What I don't like here, that the value of a capturing group should be the string instead of the subgroups.

**Comment by inf3rno** [comment]
```js
{
    value: "abc123",
    groups: [
        {
            name: "x", value: "abc", 
            groups: [
                {name: "wrd", value: "abc"}
            ]
        },
        {
            name: "y", value: "123", 
            groups: [
                {name: "num", value: "123"}
            ]
        }
    ]
}
```

**Comment by inf3rno** [comment]
I think the upper is perfect and explicit. We could simplify it this way:

```js
{
    value: "abc123",
    groups: {
        x: {
            value: "abc", 
            groups: {
                wrd: "abc"
            }
        },
        y: {
            value: "123", 
            groups: {
                num: "123"
            }
        }
    }
}
```
but we would end up checking `if num in groups`.

We could simplify it further by adding reserved word for `value`, which I think is acceptable
```js
{
    value: "abc123",
    x: {
        value: "abc", 
        wrd: "abc"
    },
    y: {
        value: "123", 
        num: "123"
    }
}
```
But again it would be a lot harder to process without knowing the exact data structure, while for the first version would could easily write a general parser. Another option to support both version somehow.

**Comment by inf3rno** [comment]
Let's check how we would process the result. The result should be `{x: "abc", y: 123}`.

**Comment by inf3rno** [comment]
```js
var o = {};
for (var group of result.groups) {
	var key = group.name;
	var subgroup = group.groups[0];
	if (subgroup.name === "num")
		o[key] = parseInt(subgroup.value);
	else
		o[key] = subgroup.value;
}
```

vs

```js
var o = {};
for (var key in result){
	if (key === "value")
		continue;
	var group = result[key];
	if (group.hasOwnProperty("num"))
		o[key] = parseInt(group.num);
	else
		o[key] = group.wrd;
}
```

**Comment by inf3rno** [comment]
In the first case the naming was a problem because we have group, subgroup, group.groups, etc... I think the `.groups` is not the proper word here, it should be `subgroups` maybe or something different e.g. `captured`, `contains`, `content`, etc... I like the subgroups though and could not come up with the proper verb for the term. I did not like the `group.groups[0]` either. We could generalize it further if we'd check the matching pattern type instead of the subgroup name, so we might be able to write a completely general reusable code for this kind of conversion.

As of the second solution, it is bad that we have to exclude the `value` key because it is reserved. We could make it non-enumerable. Another issue that in old IE the enumeration order is not necessarily the same as the property setting order, and in some cases that matters by processing the result. It is not necessary to support those browsers, but I think it is something good to have if possible. The code is a little bit longer because of the reserved word exclusion. The `group.wrd` format makes it hard to generalize it, but not completely impossible.

**Comment by inf3rno** [comment]
A general solution is this:
```js
{
    name: undefined,
    value: "abc123",
    subgroups: [
        {
            name: "x", value: "abc", pattern: x,
            subgroups: [
                {name: "wrd", value: "abc", pattern: wrd}
            ]
        },
        {
            name: "y", value: "123", pattern: y,
            subgroups: [
                {name: "num", value: "123", pattern: num}
            ]
        }
    ],
    pattern: xy
}

var o = {};
for (var group of result.subgroups) {
    var key = group.name;
    var subgroup = group.subgroups[0];
    if (subgroup.pattern instanceof NumberPattern)
        o[key] = parseInt(subgroup.value);
    else
        o[key] = subgroup.value;
}
```
We don't have to check a single group name to know how to process it. The only disadvantage that we have to keep the subpatterns in memory after compilation. To avoid that we could add the type of the subpattern instead of the subpattern itself.

**Comment by inf3rno** [comment]
```js
{
    name: undefined,
    value: "abc123",
    subgroups: [
        {
            name: "x", value: "abc", type: "alternatives",
            subgroups: [
                {name: "wrd", value: "abc", type: "word"}
            ]
        },
        {
            name: "y", value: "123", type: "alternatives",
            subgroups: [
                {name: "num", value: "123", type: "number"}
            ]
        }
    ],
    type: "concatenate"
}
```

**Comment by inf3rno** [comment]
Probably we could add a compilation flag if we want to have the exact patterns. But if so, then why do we need to name the groups with strings?

```js
var num = number();
var wrd = word();
var x = alternatives(num, wrd);
var y = copy(x);
var xy = concatenate(x,y);
var result = pattern(xy, capture(x,y,num,wrd)).match("abc123");
console.log(result);

{
    value: "abc123",
    captured: [
        {
            value: "abc", pattern: x,
            nested: [
                {value: "abc", pattern: wrd}
            ]
        },
        {
            value: "123", pattern: y,
            nested: [
                {value: "123", pattern: num}
            ]
        }
    ]
}

var o = {};
for (var group of result.captured) {
    var key = group.pattern === x ? "x" : "y";
    var subgroup = group.nested[0];
    if (subgroup.pattern === num)
        o[key] = parseInt(subgroup.value);
    else
        o[key] = subgroup.value;
}

```

**Comment by inf3rno** [comment]
I think we should have 2 different solutions here. The first solution names the patterns, the second solution does not name the patterns. If we name a pattern, then instead of the pattern value we got the name in the results. This is useful if we don't want to keep the subpatterns in memory. If we don't name the subpatterns, then we still have to describe what we want to capture.

**Comment by inf3rno** [comment]
I don't like implicit solutions, it would be far better to add flags to turn on/off subpatterns in the results:
```
{a:pa,b:pb} -> {name:"a", value: "...", pattern: pa}
[pa,pb] -> {value: "...", pattern: pa}
{a:"pa",b:"pb"} -> {name: "a", value: "...", pattern: {name:"pa",...}}
["pa","pb"] -> {value: "...", pattern: {name:"pa",...}}
```
I think this is the final version. Let's think about how to turn off subpatterns in the results.

**Comment by inf3rno** [comment]
```js
pattern(xy, capture({x:x,y:y,n:num,w:wrd}))
```
We could use options instead of calling another function for capturing groups. I did not want to use options, because I had only a single option, but now that I have 2 options we can do it.

```js
pattern(xy, {
    capture: {x:x,y:y,n:num,w:wrd},
    includeSubpatterns: true
})
```

We could convert the subpatterns e.g. `num => "num"`, but it does not make sense, since it would still keep them in the background. All we want to do is keeping the `groupIndex -> name` mapping and if it is a nested group. Everything else is wasting memory.

**Comment by inf3rno** [comment]
Do we really need these subpatterns to be able to decide howo to process something? I think we only need the capturing group names for that.

```js
var num = number();
var wrd = word();
var x = alternatives(num, wrd);
var y = copy(x);
var xy = concatenate(x,y);
var result = compile({pattern: xy, groups: {x:x,y:y,w:wrd,n:num}}).match("abc123");
console.log(result);

{
    value: "abc123", 
    subgroups: [
        {
            name: "x", value: "abc",
            subgroups: [
                {name: "w", value: "abc"}
            ]
        },
        {
            name: "y", value: "123",
            subgroups: [
                {name: "n", value: "123"}
            ]
        }
    ]
}

var o = convert(result, {
    n: function (value){
        return parseInt(value); // word is converted to process(subgroups[i])
    },
    w: function (value){
        return value; // number is converted to process(subgroups[i])
    },
    x: function (value, subvalues){
        return subvalues[0]; // alternatives is converted to [process(subgroups[i])]
    },
    y: function (value, subvalues){
        return subvalues[0];
    },
    main: function (value, subvalues){
        return subvalues; // concatenate is converted to {subgroups[i].name : process(subgroups[i])}
    }
});
console.log(o);

{
    x: "abc",
    y: 123
}

```

**Comment by inf3rno** [comment]
Conversion rules:
 - number, word -> fn(value)
 - alternatives -> fns[matchingAltName](value)
 - concatenate -> {nameForIndex(i): fn[nameForIndex(i)](value)}

**Comment by inf3rno** [comment]
What kind of assumptions do we have here?
 - there is a  named matching alternative
 - we use the same name only once in concat
 - every concatenated subpattern is in a named capturing group

**Comment by inf3rno** [comment]
Okay, check these.
 - Are all matching alternatives named? Nope, certainly not. In those cases when they aren't named we will return the raw string to the alt function instead of converting them.
 - Can we use the same name multiple times in a concat? Well in theory we could do even `x+` and capture `x` as an array instead of copying it. Or in this case we could do `x{2,2}` instead of using `y`. If we want to use this capturing system we need to deal with quantifiers at least.
 - Are all concatenated subpatterns named? Certainly not. Instead of using a map by default we need to use an array of values probably.

But all of those are about the converter and only the quantifiers are about the capturing. We can work on the converter later, for now a result with the upper format is fine. So next thing to do is adding quantifiers, backreference, etc. to the API.

**Comment by inf3rno** [comment]
We could use the same format without the `name` property:
```js
{
    value: "abc123", 
    subgroups: [
        {
            value: "abc",
            subgroups: [
                {value: "abc"}
            ]
        },
        {
            value: "123",
            subgroups: [
                {value: "123"}
            ]
        }
    ]
}
```
It is tempting to convert the `{value: "123"}` to `"123"`. It looks simpler for a human reader, but for the machine it is harder to process, because we need to add an `if (typeof(o) === "string")` condition. So I think it must stay that way. I am not sure if we should add the matching subpattern to the result.

```js
{
    value: "abc123", pattern: xy,
    subgroups: [
        {
            value: "abc", pattern: x,
            subgroups: [
                {value: "abc", pattern: wrd}
            ]
        },
        {
            value: "123", pattern: y,
            subgroups: [
                {value: "123", pattern: num}
            ]
        }
    ]
}
```

It would be fun, because we don't have to use strings to identify the groups and process the result. It is harder to process it though, because we need a HashMap for the `pattern - fn` pairs, so it is not as straightforward to convert to to the format we need as with strings. Another issue that we need to keep the subpatterns in memory and the result processing part must know of the pattern building part of the code. I think that's why it is not a good idea to support unnamed capturing groups.

**Comment by inf3rno** [comment]
The real question about this whether we really need to separate the conversion of the result from building the pattern. In most of the cases the pattern is not something standalone and we need a very specific format as a result. It was always inconvenient to process the result of normal regex because we always got an indexed array and we had to identify the groups on that. On the other hand separation of conversion makes it easier to reuse the pattern. I was wondering, is it possible to reuse a pattern without separating the result conversion from pattern building?

**Comment by inf3rno** [comment]
```js
var num = number();
var wrd = word();
var x = alternatives(num, wrd);
var y = copy(x);
var xy = concatenate(x,y);
var pattern = compile({pattern: xy, groups: {x:x,y:y,w:wrd,n:num}});
var result = pattern.match("abc123");

var o = convert(result, {
    n: function (value){
        return parseInt(value); 
    },
    w: function (value){
        return value;
    },
    x: function (value, numberOrWord){
        return numberOrWord;
    },
    y: function (value, numberOrWord){
        return numberOrWord;
    },
    main: function (value, xy){
        return xy;
    }
});
```

**Comment by inf3rno** [comment]
Solutions with subpatterns as identifiers:

```js
var num = number();
var wrd = word();
var x = alternatives(num, wrd);
var y = copy(x);
var xy = concatenate(x,y);

var pattern = compile(xy)
	.capture(num, (v) => parseInt(v))
	.capture(wrd, (v) => v)
	.capture([x,y], (v, sub) => sub[0])
	.capture(xy, (v, sub) => {x:sub[0],y:sub[1]});

var o = [];
var strings = ["abc123", "def456", "ghi789"];
for (var string of strings)
    o.push = pattern.match(string);
```

An interesting thing here, that if we don't use the subpattern in the conversion function, then we can free it up and use only its capturing group index instead. On the other hand it is harder to write the conversion this way.

**Comment by inf3rno** [comment]
In theory we could do somthing like this:

```js
pattern.convert(xy, [x,y,num,wrd], (value, sub, xyId, yId, numId, wrdId) => {
    x: sub.find(xId, or(numId, wrdId)),
    y: sub.find(yId, or(numId, wrdId))
});
```

It is a working solution, but I think it is very inconvenient.

**Comment by inf3rno** [comment]
I was thinking on giving the capturing group during subpattern definition too:

```js
var num = number().convert((v) => parseInt(v));
var wrd = word();
var x = alternatives(num, wrd);
var y = copy(x);
var xy = concatenate(x,y).convert((xv,yv) => {x:xv, y:yv};

var pattern = compile(xy);

var o = [];
var strings = ["abc123", "def456", "ghi789"];
for (var string of strings)
    o.push = pattern.match(string);
```

It is a lot easier to define anything this way, but we won't be able to reuse the subpatterns.

**Comment by inf3rno** [comment]
I need an example for how I will use this with the parser. I think that should decide which version I will keep.

**Comment by inf3rno** [comment]
Another thought, that we could use the destructuring format to describe the output we need.
```js
var num = number();
var wrd = word();
var x = alternatives(num, wrd);
var y = copy(x);
var xy = concatenate(x,y);

var pattern = compile(xy).format({x:x,y:y}).convert(num, parseInt);
```

I don't think it would be a good solution especially if we are talking about deep destructuring. It is because we have to add all the concat and alt rules when we are trying to go 1 level deeper. I would like it only because we have something similar to the output by giving the format.

**Comment by inf3rno** [comment]
I think I can accept that conversion goes the same way in most of the cases if we reuse a subpattern, but if we are talking about parsing that is not true. I think. We typically have something like `alternatives(x,y,z)` if we want to find which token we have, and if we figured out that `y` is the next, then we use the pattern of `y` with different capturing groups to process it. So the first pattern is only for detection and the second one is for processing. In the first pattern we usually don't need the capturing groups unless we are talking about backreference.

**Comment by inf3rno** [comment]
We could use copy + clear as a solution. But I don't think that is right.

**Comment by inf3rno** [comment]
Let's do this with a simple example.

```js
var a = number();
var x = concatenate(quote("x "), a);
var y = concatenate(quote("y "), a);
var t = alternatives(x,y);
var xp = compile(x);
var yp = compile(y);
var tp = compile(t);

var o = {};
var s = "x 12 y 4";
var pointer = pointer(s,0);
var r1 = tp.next(pointer);
if (r1.type == x)
    o.x = r1.value;
else if (r1.type == y)
    o.y = r1.value;
```

**Comment by inf3rno** [comment]
I think I should split up this thread.

**Comment by inf3rno** [comment]
I'll keep this thread as tracker and for brainstorming. I'll try to create a separate thread for everything I discover meanwhile.

**Comment by inf3rno** [comment]
I'll read this again and create new issues if I find something useful. I don't like tracker issues, I'd rather close this.

**Comment by inf3rno** [comment]
This thread is more about how to use transformations than a tracker thread. I'll close this and merge the transformation part with #27.
